File: src/main/java/resonantinduction/mechanical/fluid/pipe/RenderPipe.java
package resonantinduction.mechanical.fluid.pipe;

import java.util.HashMap;

import net.minecraft.util.ResourceLocation;
import net.minecraftforge.common.ForgeDirection;

import org.lwjgl.opengl.GL11;

import resonantinduction.core.Reference;
import resonantinduction.mechanical.fluid.prefab.TileFluidNetwork;

import com.builtbroken.common.Pair;

import cpw.mods.fml.client.FMLClientHandler;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class RenderPipe
{
	public static final RenderPipe INSTANCE = new RenderPipe();

	public static ModelPipe MODEL_PIPE = new ModelPipe();
	public static ModelOpenTrough MODEL_TROUGH_PIPE = new ModelOpenTrough();
	private static HashMap<Pair<EnumPipeMaterial, Integer>, ResourceLocation> TEXTURES = new HashMap<Pair<EnumPipeMaterial, Integer>, ResourceLocation>();
	public static ResourceLocation TEXTURE = new ResourceLocation(Reference.DOMAIN, Reference.MODEL_PATH + "pipe/iron.png");

	public void render(PartPipe part, double x, double y, double z, float f)
	{
		EnumPipeMaterial material = part.getMaterial();
		GL11.glPushMatrix();
		GL11.glTranslatef((float) x + 0.5F, (float) y + 1.5F, (float) z + 0.5F);
		GL11.glScalef(1.0F, -1F, -1F);
		FMLClientHandler.instance().getClient().renderEngine.bindTexture(getTexture(material, 0));
		render(material, part.getMaterialID(), part.getAllCurrentConnections());
		GL11.glPopMatrix();
	}

	public static ResourceLocation getTexture(EnumPipeMaterial mat, int pipeID)
	{
		if (mat != null)
		{
			Pair<EnumPipeMaterial, Integer> index = new Pair<EnumPipeMaterial, Integer>(mat, pipeID);

			if (!TEXTURES.containsKey(index))
			{
				String pipeName = "";
				if (EnumPipeType.get(pipeID) != null)
				{
					pipeName = EnumPipeType.get(pipeID).getName(pipeID);
				}
				TEXTURES.put(index, new ResourceLocation(Reference.DOMAIN, Reference.MODEL_PATH + "pipe/" + mat.matName + ".png"));
			}
			return TEXTURES.get(index);
		}
		return TEXTURE;
	}

	public static ResourceLocation getTexture(int meta)
	{
		return getTexture(EnumPipeMaterial.getFromItemMeta(meta), EnumPipeMaterial.getType(meta));
	}

	public static void render(EnumPipeMaterial mat, int pipeID, byte side)
	{
		if (mat == EnumPipeMaterial.WOOD)
		{
			MODEL_TROUGH_PIPE.render(side, false);
		}
		else if (mat == EnumPipeMaterial.STONE)
		{
			MODEL_TROUGH_PIPE.render(side, true);
		}
		else
		{
			if (TileFluidNetwork.canRenderSide(side, ForgeDirection.DOWN))
			{
				MODEL_PIPE.renderBottom();
			}
			if (TileFluidNetwork.canRenderSide(side, ForgeDirection.UP))
			{
				MODEL_PIPE.renderTop();
			}
			if (TileFluidNetwork.canRenderSide(side, ForgeDirection.NORTH))
			{
				MODEL_PIPE.renderBack();
			}
			if (TileFluidNetwork.canRenderSide(side, ForgeDirection.SOUTH))
			{
				MODEL_PIPE.renderFront();
			}
			if (TileFluidNetwork.canRenderSide(side, ForgeDirection.WEST))
			{
				MODEL_PIPE.renderLeft();
			}
			if (TileFluidNetwork.canRenderSide(side, ForgeDirection.EAST))
			{
				MODEL_PIPE.renderRight();
			}

			MODEL_PIPE.renderMiddle();
		}
	}

	public static void render(int meta, byte sides)
	{
		render(EnumPipeMaterial.getFromItemMeta(meta), EnumPipeMaterial.getType(meta), sides);
	}

}

File: src/main/scala/resonantinduction/mechanical/fluid/pipe/RenderPipe.scala
package resonantinduction.mechanical.fluid.pipe

import java.awt.Color

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.item.{ItemDye, ItemStack}
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.IItemRenderer
import net.minecraftforge.client.model.AdvancedModelLoader
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.fluids.FluidStack
import org.lwjgl.opengl.GL11
import resonant.content.prefab.scala.render.ISimpleItemRenderer
import resonant.lib.render.{FluidRenderUtility, RenderUtility}
import resonant.lib.utility.WorldUtility
import resonantinduction.core.Reference

@SideOnly(Side.CLIENT)
object RenderPipe extends ISimpleItemRenderer
{
  val model = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "pipe.tcn"))
  var texture = new ResourceLocation(Reference.domain, Reference.modelPath + "pipe.png")

  /**
   * Render Pipe Model
   */
  def render(meta: Int, colorCode: Int, sides: Int)
  {
    RenderUtility.enableBlending()
    RenderUtility.bind(texture)
    val material: PipeMaterials.PipeMaterial = PipeMaterials.apply(meta).asInstanceOf[PipeMaterials.PipeMaterial]
    val matColor: Color = new Color(material.color)
    GL11.glColor4f(matColor.getRed / 255f, matColor.getGreen / 255f, matColor.getBlue / 255f, 1)

    model.renderOnly("Mid")

    for (dir <- ForgeDirection.VALID_DIRECTIONS)
    {
      if (WorldUtility.isEnabledSide(sides, dir))
      {
        GL11.glColor4f(matColor.getRed / 255f, matColor.getGreen / 255f, matColor.getBlue / 255f, 1)
        var prefix: String = null

        dir match
        {
          case ForgeDirection.DOWN =>
            prefix = "Bottom"
          case ForgeDirection.UP =>
            prefix = "Top"
          case ForgeDirection.NORTH =>
            prefix = "Front"
          case ForgeDirection.SOUTH =>
            prefix = "Back"
          case ForgeDirection.WEST =>
            prefix = "Right"
          case ForgeDirection.EAST =>
            prefix = "Left"
          case _ =>
        }

        model.renderOnly(prefix + "Inter", prefix + "Connect")

        if (colorCode > 0)
        {
          val color = new Color(colorCode)
          GL11.glColor4f(color.getRed / 255f, color.getGreen / 255f, color.getBlue / 255f, 1)
        }

        model.renderOnly(prefix + "Pipe")
      }
    }

    RenderUtility.disableBlending()
  }

  /**
   * Render pipe in world
   */
  def render(part: PartPipe, x: Double, y: Double, z: Double, f: Float)
  {
    GL11.glPushMatrix()
    GL11.glTranslated(x + 0.5, y + 0.5, z + 0.5)
    render(part.getMaterialID, if (part.getColor > 0) ItemDye.field_150922_c(part.getColor) else -1, part.connectionMask)
    GL11.glPopMatrix()
    GL11.glPushMatrix()
    val fluid: FluidStack = part.tank.getFluid
    val renderSides: Int = part.connectionMask

    if (fluid != null && fluid.amount > 0)
    {
      GL11.glScaled(0.99, 0.99, 0.99)
      val tank = part.tank
      val percentageFilled: Double = tank.getFluidAmount.asInstanceOf[Double] / tank.getCapacity.asInstanceOf[Double]
      val ySouthEast = percentageFilled
      val yNorthEast = percentageFilled
      val ySouthWest = percentageFilled
      val yNorthWest = percentageFilled
      FluidRenderUtility.renderFluidTesselation(tank, ySouthEast, yNorthEast, ySouthWest, yNorthWest)
    }

    GL11.glPopMatrix()
  }

  /**
   * Render inventory pipe
   */
  def renderInventoryItem(`type`: IItemRenderer.ItemRenderType, itemStack: ItemStack, data: AnyRef*)
  {
    GL11.glPushMatrix()
    render(itemStack.getItemDamage, -1, 0xC)
    GL11.glPopMatrix()
  }
}